Add array initialisation support in default goto-harness.

NlightNFotis opened the pull request:
Joint work by me and @hannes-steffenhagen-diffblue on adding support for statically and dynamically sized arrays non-deterministic initialisation on the `function-call` harness in the new goto-harness tool.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

NlightNFotis committed:
Split up argument declaration and function call

for the function we are calling in goto-harness. Before this
we had a function called setup_parameters_and_call_entry_function,
and now we have split it up into declare_arguments and call_function.
This is so we can do some handling between declaring arguments
and calling the function.

Co-authored-by: Fotis Koutoulakis <fotis.koutoulakis@diffblue.com>
Co-authored-by: Hannes Steffenhagen <hannes.steffenhagen@diffblue.com>

NlightNFotis committed:
Add initialization of static arrays

Add non-det initialization of statically sized arrays.
This does element-wise initialisation.

Co-authored-by: Fotis Koutoulakis <fotis.koutoulakis@diffblue.com>
Co-authored-by: Hannes Steffenhagen <hannes.steffenhagen@diffblue.com>

NlightNFotis committed:
Add helper to get exactly one size value

Small helper to convert a program argument to
size_t and throw an exception if it doesn't work.

Co-authored-by: Fotis Koutoulakis <fotis.koutoulakis@diffblue.com>
Co-authored-by: Hannes Steffenhagen <hannes.steffenhagen@diffblue.com>

NlightNFotis committed:
Add dynamic array handling

Create and non-det initialise of pointers to arrays
making sure that they actually point at arrays. Also
allows binding function parameters to the size of the
array parameter. Only works with function parameters
at the moment, not with struct members or globals. Also
adds tests.

Co-authored-by: Fotis Koutoulakis <fotis.koutoulakis@diffblue.com>
Co-authored-by: Hannes Steffenhagen <hannes.steffenhagen@diffblue.com>

NlightNFotis committed:
Add recursive_initialization_configt::to_string

Added a debugging helper function. It was worth
keeping it around even after debugging.

Co-authored-by: Fotis Koutoulakis <fotis.koutoulakis@diffblue.com>
Co-authored-by: Hannes Steffenhagen <hannes.steffenhagen@diffblue.com>

hannes-steffenhagen-diffblue commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -16,8 +16,12 @@ Author: Diffblue Ltd.
 #include <util/ui_message.h>
 
+#include <algorithm>
+#include <set>
```
These includes aren't actually needed in this, but a later commit. Should be moved

hannes-steffenhagen-diffblue commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -185,6 +185,15 @@ void function_call_harness_generatort::validate_options()
+{
+  static auto const string_value = require_exactly_one_value(option, values);
+  // TODO replace this with string2optional
+  return narrow<std::size_t>(std::stoull(string_value));
```
`string2optional` is available now (we rebased on develop already) so we should use it here

NlightNFotis replied:
:heavy_check_mark:

hannes-steffenhagen-diffblue commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -42,6 +43,12 @@ struct function_call_harness_generatort::implt
   std::unique_ptr<recursive_initializationt> recursive_initialization;
 
+  std::set<irep_idt> function_parameters_to_treat_as_arrays;
+  std::set<irep_idt> function_arguments_to_treat_as_arrays;
```
I don't particularly enjoy the solution we have here - basically, we need to keep track of the names of function _arguments_ as declared on the function, and their mapping to _parameters_, and which _parameters_ the array initialisation applies to etc...

NlightNFotis replied:
This is just a note, we don't need to do it right now.

hannes-steffenhagen-diffblue commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -144,6 +206,17 @@ void function_call_harness_generatort::implt::generate(
   auto const arguments = declare_arguments(function_body);
 
   recursive_initialization_config.mode = function_to_call.mode;
+  recursive_initialization_config.pointers_to_treat_as_arrays =
```
This part is getting a bit more involved for one function than I am comfortable with, may be worth splitting this up into another function

NlightNFotis replied:
We decided to leave it as is right now, but to add an extra comment for slightly better documentation.

hannes-steffenhagen-diffblue commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -37,7 +38,16 @@ void recursive_initializationt::initialize(
   else if(type.id() == ID_pointer)
   {
-    initialize_pointer(lhs, depth, known_tags, body);
+    if(
```
This could have been solved more elegantly if we used a handler chain pattern as the test-gen team does for their harness generation. Didn't want to do this atm because this was mostly about porting existing functionality, but might be worth converting this at some point.

NlightNFotis replied:
This is just a note, we don't need to do it right now.

hannes-steffenhagen-diffblue reviewed (commented):
We'll also need to update user documentation ideally with usage examples and current limitations (no support for globals and struct members for dynamic arrays atm)

xbauch reviewed (approved):
Looks good.

sonodtt commented:
> We'll also need to update user documentation ideally with usage examples and current limitations (no support for globals and struct members for dynamic arrays atm)

Superb. The less time anyone needs to spend 'reverse-engineering' context/boundary conditions/ edge case coverage - is all company time that can be used for other purposes. A huge win at little cost. 
To some extent - the same applies to a PR - in general the less assumptions made about people knowing what is in/ "the purpose" / "edge conditions" of a PR the better. I recall also external contributors making the same point.

I am delighted to see that you're keeping this in mind (& I know Fotis also shares this perspective).

NlightNFotis committed:
Add goto-harness documentation to cprover-manual

This adds some user documentation to goto-harness. Currently only includes
documentation for the call-function harness, not the memory-snapshot harness

hannes-steffenhagen-diffblue commented:
Note that some of the examples don’t quite work atm due to a bug that has a pending fix in another PR 

https://github.com/diffblue/cbmc/pull/4277

tautschnig commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -56,6 +54,11 @@ struct function_call_harness_generatort::implt
+  code_function_callt::argumentst declare_arguments(code_blockt &function_body);
+  void call_function(
+    const code_function_callt::argumentst &arguments,
+    code_blockt &function_body);
```
Maybe add documentation? The previous method had one ;-)

tautschnig commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -266,3 +240,44 @@ void function_call_harness_generatort::implt::
+  for(const auto &parameter : parameters)
+  {
+    auto argument = allocate_objects.allocate_automatic_local_object(
+      parameter.type(), parameter.get_base_name());
```
Isn't use of the base_name dangerous? (It's really just something that's easy to read, but never to be used as an identifier - unless this code creates some fresh identifier out of it?)

hannes-steffenhagen-diffblue replied:
It’s used to synthesise a new variable with a similar name here

tautschnig commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -129,3 +134,23 @@ void recursive_initializationt::initialize_nondet(
+  PRECONDITION(array.type().id() == ID_array);
+  const auto &array_type = to_array_type(array.type());
+  const auto array_size =
+    numeric_cast_v<std::size_t>(to_constant_expr(array_type.size()));
```
Are you sure this works, i.e., will the caller only ever pass in arrays that have a constant size?

hannes-steffenhagen-diffblue replied:
As of right now, yes. Atm we only initialise global variables and function parameters, VLAs are not allowed in either of these contexts. Extending this to work with VLAs would be a bit tricky to be honest, because we unroll loops statically. The only reason we do this though is because we ran into a number of bugs trying to make this work with dynamic allocation, once those are fixed we can just use dynamic allocation and leave the unrolling of loops to cbmc, which should make this more flexible and the generated code would become a lot more readable as a bonus.

tautschnig replied:
Could you please add a comment to the code, pretty much stating what you just said? It seems perfectly possible that this invariant fails at some point, and then the person doing the debugging should have all the information immediately available.

tautschnig commented on src/goto-harness/recursive_initialization.h:
```diff
@@ -69,6 +69,13 @@ class recursive_initializationt
     const recursion_sett &known_tags,
     code_blockt &body);
+  // array stuff
+  // static array handling
```
I'm not sure I'm able to make much sense of these comments?! (Certainly not more than what the pretty descriptive name "initialize_array" provides as information.)

tautschnig commented on regression/goto-harness/pointer-to-array-function-parameters-min-size/test.desc:
```diff
@@ -0,0 +1,8 @@
+--harness-type call-function --function min_array_size_test --max-array-size 3 --min-array-size 3 --associated-array-size arr:sz
+VERIFICATION SUCCESSFUL
+EXIT=0
+SIGNAL=0
```
`^EXIT=0$`, `^SIGNAL=0$`

tautschnig commented on regression/goto-harness/pointer-to-array-function-parameters-multi-arg-right/test.desc:
```diff
@@ -0,0 +1,8 @@
+test.c
+--harness-type call-function --function is_prefix_of --treat-pointer-as-array string --treat-pointer-as-array prefix --associated-array-size string:string_size --associated-array-size prefix:prefix_size --max-array-size 5
+SIGNAL=0
+EXIT=0
```
anchoring as above

tautschnig commented on regression/goto-harness/pointer-to-array-function-parameters-multi-arg-wrong/test.desc:
```diff
@@ -0,0 +1,7 @@
+test.c
+--harness-type call-function --function is_prefix_of --treat-pointer-as-array string --treat-pointer-as-array prefix --associated-array-size string:string_size --associated-array-size prefix:prefix_size --max-array-size 5
+EXIT=10
+SIGNAL=0
```
anchoring

NlightNFotis replied:
All of these got handled at the latest push :)

tautschnig commented on regression/goto-harness/pointer-to-array-function-parameters-with-size/test.desc:
```diff
@@ -0,0 +1,11 @@
+test.c
+--harness-type call-function --function test --treat-pointer-as-array arr --associated-array-size arr:sz
+EXIT=0
+SIGNAL=0
```
anchoring

tautschnig commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -114,6 +124,61 @@ void function_call_harness_generatort::handle_option(
+    for(auto const &array_size_pair : values)
+    {
+      std::string array;
+      std::string size;
```
Why declare them outside the `try` block?

hannes-steffenhagen-diffblue replied:
Just an oversight I think

tautschnig commented on src/goto-harness/recursive_initialization.h:
```diff
@@ -10,17 +10,28 @@ Author: Diffblue Ltd.
 
 #include <goto-programs/goto_model.h>
 #include <util/expr.h>
+#include <util/irep.h>
```
You surely already have `irep.h` if you include `expr.h`

tautschnig commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -0,0 +1,245 @@
+* `function` is the name of the function that gets instrumented
+* then we pass the input GOTO-binary and a name for the output GOTO binary.
+
+What comes of this is a GOTO binary the following C pseudocode:
```
Missing some word - maybe "... binary _implementing_ the ..."?

tautschnig commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -0,0 +1,245 @@
+
+We also have support for arrays (currently only for array function parameters, globals and struct members are considered for the future).
+
+Take this example of a an implementation of an `is_prefix_of` function that should return true if the first string parameter `string` is a prefix of the second one `prefix`.
```
"a an"

tautschnig commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -0,0 +1,245 @@
+`harness` as the entry function for analysis, or further modify
+it with `goto-harness` or `goto-instrument` or other tools.
+
+The example above demonstrates the most simple case, which is roughly the same as the entry point `cbmc` automically generates for functions. However, the `function-call` harness can also non-deterministically
```
There are some very long lines here...

tautschnig commented on doc/cprover-manual/goto-harness.md:
````diff
@@ -0,0 +1,245 @@
+$ cbmc --function harness --pointer-check array_example-mod.gb
+```
+
+We have the new option `associated-array-size` here. This is in the format `<array-parameter-name>:<array-size-parameter-name>` and will cause the array parameter with name`array-parameter-name` to be initialised as an array, with the parameter `array-size-parameter-name` should hold its size (it should have some integral type like `int` or `size_t`).
````
s/new option/additional option/, s/should hold its/holding its/

tautschnig commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -0,0 +1,245 @@
+
+By default, arrays are created with a minimum size of 1 and a maximum size of 2. These limits can be set with the `--min-array-size` and `--max-array-size` options.
+
+If you have a function that takes and array parameter, but without an associated size parameter, you can also use the `--treat-pointer-as-array <parameter-name>` option.
```
s/and array/an array/

tautschnig reviewed (approved):
A couple of comments, but also a rebase and making-CI-happy is required.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 0ad43fd848e9d3a7e1dd11968550a1c74432a9e7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102341655
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 151dfd2655c4c376016730de3a0d69192ea67791).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102349568
Status will be re-evaluated on next push.
Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

  - the compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 07c31d7b829e0446e6b6e7b3fcce40502fe2e2cc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102353668

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 5904fec4294ac3306456f4d5079107e6fce5afa7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102477427

xbauch commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -0,0 +1,263 @@
+* The `function-call` harness, which automatically synthesises an environment
+without having any information about the program state.
+
+It is used similar to how `goto-instrument` is used by modifying an existing
```
similarly

xbauch commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -0,0 +1,263 @@
+The example above demonstrates the most simple case, which is roughly the same
+as the entry point `cbmc` automically generates for functions. However, the
+`function-call` harness can also non-deterministically initialise global
+variables, array and struct elements, consider this more complicated example:
```
elements. Consider

xbauch commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -0,0 +1,263 @@
+$ goto-harness \
+   --harness-function-name harness \
+   --harness-type call-function \
+   --function check_list
```
Shouldn't there also be the `\` at the end of the line?

xbauch commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -0,0 +1,263 @@
+   --nondet-globals \
+   list_example.gb \
+   list_example_with_harness.gb
+$ cbmc --function harness list_example_with_harness.gb --unwind 20 --unwinding-assertions
```
Do we really need to bound the unwinding?

hannes-steffenhagen-diffblue replied:
Yes, unfortunately symex can’t seem to figure out when to stop unwinding on its own here,

LAJW commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -136,5 +151,180 @@ void recursive_initializationt::initialize_nondet(
+  const irep_idt &array_name) const
+{
+  return initialization_config.pointers_to_treat_as_arrays.find(array_name) !=
+         initialization_config.pointers_to_treat_as_arrays.end();
```
⛏ `set::count` is shorter. You could even get away without this method - especially since it's used only once.
```c++
initialization_config.pointers_to_treat_as_arrays.count(array_name);
```

hannes-steffenhagen-diffblue replied:
The reason these are in named functions is twofold:
1. The logic for these will get more complicated in the future if we decide to support more complex scenarios
2. This is spelling out what it's intended to do, rather than just showing what it _does_ do

LAJW replied:
You still can use `count` though.

LAJW commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -136,5 +151,180 @@ void recursive_initializationt::initialize_nondet(
+  code_blockt &body)
+{
+  PRECONDITION(array.type().id() == ID_array);
+  const auto &array_type = to_array_type(array.type());
```
⛏ I think you should change the type of `array` to `array_exprt` - that way you wouldn't have to put these invariants here.

tautschnig replied:
I don't think that works here: an `array_exprt` is not the same as an `array_typet`-typed expression. (An `array_exprt` is an array, while an `array_typet`-typed expression could as well just be a symbol.)

LAJW commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -136,5 +151,180 @@ void recursive_initializationt::initialize_nondet(
+
+void recursive_initializationt::initialize_array(
+  const exprt &array,
+  std::size_t depth,
```
:pick: `const`.

LAJW commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -111,6 +127,61 @@ void function_call_harness_generatort::handle_option(
+      values.end(),
+      std::inserter(
+        p_impl->function_parameters_to_treat_as_arrays,
+        p_impl->function_parameters_to_treat_as_arrays.end()));
```
⛏ `set` has a 2-argument insert method which tends to be more efficient and happens to be shorter.
```c++
p_impl->function_parameters_to_treat_as_arrays.insert(values.begin(), values.end());
```

LAJW commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -40,14 +45,17 @@ struct function_call_harness_generatort::implt
   std::unique_ptr<recursive_initializationt> recursive_initialization;
 
+  std::set<irep_idt> function_parameters_to_treat_as_arrays;
+  std::set<irep_idt> function_arguments_to_treat_as_arrays;
```
❓ Why `std::set` instead of `std::unordered_set`. Given what you're doing the former would be much more efficient.

hannes-steffenhagen-diffblue replied:
In practical terms I don't expect these will ever contain more than _maybe_ a couple dozen or so members, so don't know about _much_ more efficient. But it's not like changing to `unordered_` is going to break anything, so I don't mind it either.

LAJW commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -111,6 +127,61 @@ void function_call_harness_generatort::handle_option(
+        p_impl->function_parameters_to_treat_as_arrays.insert(array);
+        auto const inserted =
+          p_impl->function_parameter_to_associated_array_size.insert(
+            {array, size});
```
⛏ `emplace` will allow you to avoid the temporary (and get rid of curly braces):
```c++
p_impl->function_parameter_to_associated_array_size.emplace(array, size);
```

hannes-steffenhagen-diffblue replied:
I do recall having problems with emplace support on VS a while ago. Maybe that has changed since we moved away from AppVeyor?

tautschnig replied:
Yes, I think `emplace` works fine now.

LAJW commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -125,55 +196,38 @@ void function_call_harness_generatort::generate(
+    std::inserter(
+      recursive_initialization_config.variables_that_hold_array_sizes,
+      recursive_initialization_config.variables_that_hold_array_sizes.end()),
+    [](const std::pair<irep_idt, irep_idt> &p) { return p.second; });
```
🤔 In trivial cases like this a for-loop might be clearer:
```c++
for (const auto &pair : function_argument_to_associated_array_size)
  recursive_initialization_config.variables_that_hold_array_sizes(pair.second);
```
These names are insane BTW.

LAJW commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -220,6 +274,32 @@ void function_call_harness_generatort::validate_options()
+  const std::string &option,
+  const std::list<std::string> &values)
+{
+  static auto const string_value = require_exactly_one_value(option, values);
```
🚫 argument-dependent `static`?! What's going on here?

hannes-steffenhagen-diffblue replied:
Yep, that's definitely not intended

NlightNFotis replied:
good catch, don't know how we did this.

LAJW commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -220,6 +274,32 @@ void function_call_harness_generatort::validate_options()
+{
+  static auto const string_value = require_exactly_one_value(option, values);
+  auto value = string2optional<std::size_t>(string_value, 10);
+  if(value.has_value())
```
⛏ You can inline the statement and use the following syntax (for which optionals were designed):
```c++
if(auto const value = string2optional<std::size_t>(string_value, 10))
```

NlightNFotis replied:
No. I don't ever want to assign a value in an if condition.

LAJW commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -278,3 +358,74 @@ void function_call_harness_generatort::implt::
+    auto const &argument_name = p.second;
+    if(
+      function_parameters_to_treat_as_arrays.find(parameter_name) !=
+      function_parameters_to_treat_as_arrays.end())
```
⛏ `std::set::count` to keep it more DRY.

tautschnig replied:
If so, please use `if(function_parameters_to_treat_as_arrays.count(parameter_name) != 0)` to keep Visual Studio happy.

hannes-steffenhagen-diffblue replied:
I don't know in what way that's more DRY, but it's certainly shorter so sure.

LAJW commented on src/goto-harness/function_call_harness_generator.cpp:
```diff
@@ -278,3 +358,74 @@ void function_call_harness_generatort::implt::
+    arguments.push_back(argument);
+  }
+
+  for(const auto &p : parameter_name_to_argument_name)
```
⛏ Why not call it a `pair`? it wouldn't cause line wrap. Not using abbreviations is in our coding standard after all.

LAJW commented on src/goto-harness/recursive_initialization.h:
```diff
@@ -69,6 +81,21 @@ class recursive_initializationt
+  bool should_be_treated_as_array(const irep_idt &pointer_name) const;
+  bool is_array_size_parameter(const irep_idt &cmdline_arg) const;
+  optionalt<irep_idt>
+  get_associated_size_variable(const irep_idt &array_name) const;
```
⛏ These three functions are wrappers around one-liners used only once, I think they should be inlined.

LAJW commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -136,5 +151,180 @@ void recursive_initializationt::initialize_nondet(
+  auto min_array_size = initialization_config.min_dynamic_array_size;
+  auto max_array_size = initialization_config.max_dynamic_array_size;
+  PRECONDITION(max_array_size >= min_array_size);
+  std::vector<symbol_exprt> array_variables;
```
💡 I think you should group creation of these locals into immediately-invoked lambda expressions (at the very least). As it is, this function is very difficult to follow, it's very long, everything is mutable and potentially touched.

hannes-steffenhagen-diffblue replied:
There are definitely a couple places where variables that are meant to be `const` aren't marked as such and should be fixed. I'll take a look at how the function would look with immediately invoked lambdas to see if that'd actually improve the readability.

LAJW commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -136,5 +151,180 @@ void recursive_initializationt::initialize_nondet(
+    array_name);
+}
+
+void recursive_initializationt::initialize_dynamic_array(
```
It looks like `initialization_config` and `goto_model` are the only member variables used in this function.

🤔 I wonder if it could be made free.

NlightNFotis replied:
Personally not fans of "free" functions, I don't consider them a sign of good design. Not going to do that, sorry.

hannes-steffenhagen-diffblue replied:
@LAJW You can make any function free, I'm just not sure what the point would be in this instance.

LAJW replied:
As it is, this code tells me the entire class is necessary (which is not true - only 2 members are).

If you were to make it free, it would be unit-testable. And possible to reuse elsewhere. This code looks kind of generic. I think it's been done before.

If it was not bound to this particular context, it could be reused. But because it's a private method that adds something to a private field, that's not possible. So I guess that'll mean more comments like the one with which this function starts.

NlightNFotis replied:
As it is right now, this is a method of the recursive initialiser class, which is also tasked with initialising other things as well (pointers, strings,etc). If this were a free function, then it would be isolated it from its logical context, and not make a lot of sense, just to satisfy blindly a random design guideline which is not very appropriate in this case.

Let's please not look at the trees and lose sight of the forest.

hannes-steffenhagen-diffblue replied:
> As it is, this code tells me the entire class is necessary (which is not true - only 2 members are).

Basically I've been thinking about an approach to clean this up somewhat, however this would require significant rewrites of the entire functionality we have here so far, which I don't think makes sense to do as part of this PR.

LAJW commented on src/goto-harness/recursive_initialization.cpp:
```diff
@@ -136,5 +151,180 @@ void recursive_initializationt::initialize_nondet(
+  const exprt &pointer,
+  std::size_t depth,
+  const recursion_sett &known_tags,
+  code_blockt &body)
```
❓ Why have an `out` argument, when you could return the list of statements to add.

NlightNFotis replied:
It's a design decision. We could go that way, for sure, but in our initial design of goto-harness we went this way, and now we want to keep it for interface uniformity.

LAJW replied:
But this is an implementation function - not a part of the interface, so you're not limited in that regard. 🤔

LAJW reviewed (commented):
Rough.

That random argument-dependent-static variable is VERY suspicious.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 1be00d657f5fcf43a45adfdad1ff09a8757ee6e5).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102514471

chrisr-diffblue commented on regression/goto-harness/chain.sh:
```diff
@@ -23,5 +23,11 @@ if [ -e "${name}-mod.gb" ] ; then
 
-$goto_harness "${name}.gb" "${name}-mod.gb" --harness-function-name $entry_point ${args} 
-$cbmc --function $entry_point "${name}-mod.gb" --unwind 20 --unwinding-assertions
+# `# some comment` is an inline comment - basically, cause bash to execute an empty command
```
Nice!

chrisr-diffblue commented on src/goto-harness/recursive_initialization.h:
```diff
@@ -31,6 +31,8 @@ struct recursive_initialization_configt
   std::set<irep_idt> variables_that_hold_array_sizes;
   std::map<irep_idt, irep_idt> array_name_to_associated_array_size_variable;
+
+  std::string to_string() const; // for debugging purposes
```
If `to_string()` is only meant to be used for debugging purposes, is it worth wrapping it in `#if DEBUG` ?

chrisr-diffblue commented on doc/cprover-manual/goto-harness.md:
```diff
@@ -0,0 +1,263 @@
+
+We have two types of harnesses we can generate for now:
+
+* The `memory-snapshot` harness. TODO: Daniel can document this.
```
There's more than one 'Daniel'....

tautschnig replied:
Maybe explicitly reference the PR (#4261)?

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 0ea5c31e6e7154da7a8f8880ed2a26f5d87801a9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/102542412
